### PR TITLE
Checking for returned errors when updating conditions & finalizers

### PR DESCRIPTION
**Pull request by arschles**
This patch adds error checks to all `update{Broker,Binding,Instance}Condition` calls in the controller, along with tests to ensure that the errors are properly handled.

Fixes https://github.com/kubernetes-incubator/service-catalog/issues/925

### CODE REVIEW

**Comment by arschles** [comment]
This is a work in progress. More changes and (likely) some tests to come

**Comment by arschles** [comment]
There are lots more test cases to check for each edge case (every place where `update*Condition` might return an error) but I'd like to get this change in and do those in a follow-up. I'm taking off the `in-progress` label to solicit reviews.

cc/ @vaikas-google @pmorie @krancour

File: pkg/controller/controller_binding.go [hunk 1]
```diff
@@ -119,13 +122,16 @@ func (c *controller) reconcileBinding(binding *v1alpha1.Binding) error {
...
 			v1alpha1.BindingConditionReady,
 			v1alpha1.ConditionFalse,
 			errorWithOngoingAsyncOperation,
 			errorWithOngoingAsyncOperationMessage,
-		)
+		); err != nil {
+			glog.Errorf("updating binding condition (%s)", err)
+			return err
+		}
 		c.recorder.Event(binding, api.EventTypeWarning, errorWithOngoingAsyncOperation, s)
```
Comment (MHBauer) [inline pkg/controller/controller_binding.go]: do we want to put these events above the binding updates? we'll skip them now.

Comment (arschles) [inline pkg/controller/controller_binding.go]: @MHBauer I left them where they were before. I think it's the right behavior to not emit an event if the update failed, because the state didn't change. Thoughts?

Comment (MHBauer) [inline pkg/controller/controller_binding.go]: What causes updating the condition in status to fail?

What happens after? Do we try to reconcile again? Eventually it succeeds?

File: pkg/controller/controller_binding.go [hunk 2]
```diff
@@ -216,36 +234,46 @@ func (c *controller) reconcileBinding(binding *v1alpha1.Binding) error {
...
-			)
+			); err != nil {
+				glog.Errorf("updating binding condition (%s)", err)
+				return err
+			}
 			c.recorder.Event(binding, api.EventTypeWarning, errorInjectingBindResultReason, s)
 			return err
 		}
-		c.updateBindingCondition(
+		if err := c.updateBindingCondition(
```
Comment (vaikas-google) [inline pkg/controller/controller_binding.go]: I'm finding this style here to be harder to read than (IMHO) simpler:
err := c.update...
if err != nil {


But, if others are fine with this style, then so be it :)

File: pkg/controller/controller_binding.go [hunk 3]
```diff
@@ -289,26 +320,36 @@ func (c *controller) reconcileBinding(binding *v1alpha1.Binding) error {
...
 			"The binding was deleted successfully",
-		)
+		); err != nil {
+			glog.Errorf("updating binding condition (%s)", err)
+			return err
+		}
 		// Clear the finalizer
 		finalizers.Delete(v1alpha1.FinalizerServiceCatalog)
-		c.updateBindingFinalizers(binding, finalizers.List())
+		if err := c.updateBindingFinalizers(binding, finalizers.List()); err != nil {
```
Comment (vaikas-google) [inline pkg/controller/controller_binding.go]: So for example here, I think this is more readable using this format because things can fit in a single line :)

File: pkg/controller/controller_binding.go [hunk 4]
```diff
@@ -119,13 +122,16 @@ func (c *controller) reconcileBinding(binding *v1alpha1.Binding) error {
...
-		c.updateBindingCondition(
+		if err := c.updateBindingCondition(
 			binding,
 			v1alpha1.BindingConditionReady,
 			v1alpha1.ConditionFalse,
 			errorWithOngoingAsyncOperation,
 			errorWithOngoingAsyncOperationMessage,
-		)
+		); err != nil {
+			glog.Errorf("updating binding condition (%s)", err)
```
Comment (vaikas-google) [inline pkg/controller/controller_binding.go]: Should we be creating an error event in these failure cases?

Comment (vaikas-google) [inline pkg/controller/controller_binding.go]: I mean error about the update condition failing?

File: pkg/controller/controller_binding.go (hunk 1 above)
Comment (pmorie) [inline pkg/controller/controller_binding.go]: > What causes updating the condition in status to fail?

Broken pipe, API server goes down, etc

> What happens after? Do we try to reconcile again? Eventually it succeeds?

We periodically re-reconcile everything, yeah.  Eventually it succeeds.

File: pkg/controller/controller_binding.go [hunk 5]
```diff
@@ -164,13 +173,16 @@ func (c *controller) reconcileBinding(binding *v1alpha1.Binding) error {
...
-				c.updateBindingCondition(
+				if err := c.updateBindingCondition(
 					binding,
 					v1alpha1.BindingConditionReady,
 					v1alpha1.ConditionFalse,
 					errorWithParameters,
 					"Error unmarshaling binding parameters. "+s,
-				)
+				); err != nil {
+					glog.Errorf("updating binding condition (%s)", err)
```
Comment (pmorie) [inline pkg/controller/controller_binding.go]: I would prefer to include the text "Error" in the message, even though I am fully aware that the log level says it's an error.

File: pkg/controller/controller_reconcile_binding_table_test.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+package controller
+
+import (
+	scfake "github.com/kubernetes-incubator/service-catalog/pkg/client/clientset_generated/clientset/fake"
+	clientgotesting "k8s.io/client-go/testing"
+)
+
+// bindingTestCase represents a single row of a table driven test for reconciling bindings
+type reconcileBindingTestCase struct {
```
Comment (pmorie) [inline pkg/controller/controller_reconcile_binding_table_test.go]: strongly prefer for this to be defined where it is used

File: pkg/controller/controller_binding.go [hunk 7]
```diff
@@ -216,36 +234,46 @@ func (c *controller) reconcileBinding(binding *v1alpha1.Binding) error {
...
 			glog.Warning(s)
-			c.updateBindingCondition(
+			if err := c.updateBindingCondition(
 				binding,
 				v1alpha1.BindingConditionReady,
 				v1alpha1.ConditionFalse,
 				errorBindCallReason,
-				"Bind call failed. "+s)
+				"Bind call failed. "+s,
+			); err != nil {
```
Comment (MHBauer) [inline pkg/controller/controller_binding.go]: note to self, feel free to read and confirm cases : binding creation failed, can't update condition to say binding failed. next time through loop will try to create binding again.

File: pkg/controller/controller_binding.go [hunk 8]
```diff
@@ -216,36 +234,46 @@ func (c *controller) reconcileBinding(binding *v1alpha1.Binding) error {
...
-		c.updateBindingCondition(
+		if err := c.updateBindingCondition(
 			binding,
 			v1alpha1.BindingConditionReady,
 			v1alpha1.ConditionTrue,
 			successInjectedBindResultReason,
 			successInjectedBindResultMessage,
-		)
+		); err != nil {
+			glog.Errorf("updating binding condition (%s)", err)
```
Comment (MHBauer) [inline pkg/controller/controller_binding.go]: binding successfully injected,  next time through the loop we will try to create and inject a fresh binding?

This seems off to me.

File: pkg/controller/controller_binding.go [hunk 9]
```diff
@@ -266,13 +294,16 @@ func (c *controller) reconcileBinding(binding *v1alpha1.Binding) error {
...
-			c.updateBindingCondition(
+			if err := c.updateBindingCondition(
 				binding,
 				v1alpha1.BindingConditionReady,
 				v1alpha1.ConditionUnknown,
 				errorEjectingBindReason,
 				errorEjectingBindMessage+s,
-			)
+			); err != nil {
+				glog.Errorf("updating binding condition (%s)", err)
```
Comment (MHBauer) [inline pkg/controller/controller_binding.go]: if we fail below, do we come back through here and fail to eject the binding? (as it is already ejected?) 

How does  ConditionUnknown play into this?

File: pkg/controller/controller_reconcile_binding_table_test.go (hunk 6 above)
Comment (MHBauer) [inline pkg/controller/controller_reconcile_binding_table_test.go]: Why is this a separate file?

**Comment by arschles** [comment]
This is a relatively complex rebase, and is ongoing. I will update status here when it's ready for review again.

**Comment by duglin** [comment]
just a reminder of a rebase needed.

**Comment by pmorie** [comment]
@arschles build failure:

<details>

```
Running gofmt:
*** Please gofmt the following:
pkg/controller/controller_binding_test.go:84:4: expected declaration, found ')'
```

</details>

**Comment by kibbles-n-bytes** [comment]
Also a non-lint error:

```[Build & Unit/Integration Tests] pkg/controller/controller_instance.go:549: undefined: toUpdae```

**Comment by staebler** [comment]
Agree with @n3wscott. This PR has been obviated by #1149.

**Comment by pmorie** [comment]
Closing since this is obviously stale.
